### PR TITLE
RavenDB-19298 Helm: Install via artifacthub.io

### PR DESCRIPTION
**Pull request by poissoncorp**
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-19298/Helm-Install-via-artifacthubio

### Additional description

Initial patch of the `rvn create-setup-package` that allows using `--generate-helm-values|-g [path]` option

### Type of change

- New feature

### How risky is the change?

- Low 

### Backward compatibility

- Non breaking change

### Documentation update

- This change requires a documentation update. Please mark the issue on YouTrack using `Documentation Required` tag.

### Testing 

- It has been verified by manual testing

### Is there any existing behavior change of other features due to this change?

- Yes and no - it simplifies our Helm chart flow, I'll link a PR from helm chart repo here 
- https://github.com/ravendb/helm-charts/pull/18

### CODE REVIEW

File: tools/rvn/CommandLineApp.cs [hunk 1]
```diff
@@ -575,6 +606,11 @@ private static CommandOption ConfigureCertPassword(CommandLineApplication cmd)
             return cmd.Option("-p|--password", $"Certificate password{Environment.NewLine}Password can be set from ENV:{Environment.NewLine}Windows - $env:RVN_CERT_PASS=password\nLinux - export RVN_CERT_PASS=password", CommandOptionType.SingleValue);
         }
 
+        private static CommandOption ConfigureGenerateValues(CommandLineApplication cmd)
+        {
+            return cmd.Option("-g|--generate-helm-values", "Path to values.yaml", CommandOptionType.SingleValue);
```
Comment (gregolsky) [inline tools/rvn/CommandLineApp.cs]: Let's use only the long option name for this one.

File: tools/rvn/CommandLineApp.cs [hunk 2]
```diff
@@ -645,6 +681,41 @@ private static void ValidateSetupOptions(CreateSetupPackageParameters parameters
...
+        {
+            using var context = JsonOperationContext.ShortTermSingleUse();
+            var jsonBlittable = context.ReadObject(parameters.SetupInfo.License.ToJson(), "license");
+            HelmInfo helmInfo = new()
+            {
+                Domain = $"{parameters.SetupInfo.Domain}.{parameters.SetupInfo.RootDomain}",
+                Email = parameters.SetupInfo.Email,
+                License = jsonBlittable.ToString(),
+                NodeTags = parameters.SetupInfo.NodeSetupInfos.Keys.ToList(),
+                SetupMode = parameters.Mode=="lets-encrypt"?"LetsEncrypt":"Secured",
```
Comment (gregolsky) [inline tools/rvn/CommandLineApp.cs]: spacing

File: tools/rvn/HelmInfo.cs [hunk 3]
```diff
@@ -0,0 +1,18 @@
+﻿using System.Collections.Generic;
+
+namespace rvn
+{
+    public class HelmInfo
+    {
+        public string StorageSize { get; set; } = "5Gi";
```
Comment (gregolsky) [inline tools/rvn/HelmInfo.cs]: Let's put 10Gi here by default

**Comment by gregolsky** [comment]
lgtm

File: tools/rvn/rvn.csproj [hunk 4]
```diff
@@ -14,8 +14,6 @@
     <ValidateExecutableReferencesMatchSelfContained>false</ValidateExecutableReferencesMatchSelfContained>
   </PropertyGroup>
   <ItemGroup>
-    <Compile Include="..\..\src\Raven.Client\Extensions\MemberInfoExtensions.cs" Link="Client\Extensions\MemberInfoExtensions.cs" />
```
Comment (gregolsky) [inline tools/rvn/rvn.csproj:17]: wasn't it needed or sth?

File: tools/rvn/CommandLineApp.cs [hunk 5]
```diff
@@ -144,6 +153,7 @@ private static void ConfigureSetupPackage()
                     var packageOutPathVal = packageOutPath.Value();
                     var certPathVal = certPath.Value();
                     var certPassTuple = certPass.Value() ?? Environment.GetEnvironmentVariable("RVN_CERT_PASS");
+                    var generateHelmValuesVal = generateHelmValues.HasValue() && generateHelmValues.Value() is null ? "values.yaml" : generateHelmValues.Value();
```
Comment (gregolsky) [inline tools/rvn/CommandLineApp.cs:156]: And if it does not have a value we'll still try to do `.Value()`?

File: tools/rvn/CommandLineApp.cs [hunk 6]
```diff
@@ -575,6 +653,11 @@ private static CommandOption ConfigureCertPassword(CommandLineApplication cmd)
             return cmd.Option("-p|--password", $"Certificate password{Environment.NewLine}Password can be set from ENV:{Environment.NewLine}Windows - $env:RVN_CERT_PASS=password\nLinux - export RVN_CERT_PASS=password", CommandOptionType.SingleValue);
         }
 
+        private static CommandOption ConfigureGenerateValues(CommandLineApplication cmd)
+        {
+            return cmd.Option("--generate-helm-values", "Path to values.yaml", CommandOptionType.SingleOrNoValue);
```
Comment (gregolsky) [inline tools/rvn/CommandLineApp.cs]: Doesn't this method have a parameter for supplying default value?
Or a property after it's created?

File: tools/rvn/CommandLineApp.cs [hunk 7]
```diff
@@ -645,6 +728,41 @@ private static void ValidateSetupOptions(CreateSetupPackageParameters parameters
...
+            if (string.IsNullOrWhiteSpace(parameters.HelmValuesOutputPath))
+            {
+                throw new InvalidOperationException("Please provide a valid file name for the helm values yaml.");
+            }
+            
+            if (Path.HasExtension(parameters.HelmValuesOutputPath) == false)
+            {
+                parameters.HelmValuesOutputPath += ".yaml";
+            }
+            else if (Path.GetExtension(parameters.HelmValuesOutputPath)?.Equals(".yaml", StringComparison.OrdinalIgnoreCase) == false)
```
Comment (gregolsky) [inline tools/rvn/CommandLineApp.cs:745]: Are you sure Path.GetExtension() is not going to return null? Maybe look for something more bulletproof.

File: src/Raven.Server/Commercial/SetupWizard/LetsEncryptSetupUtils.cs [hunk 8]
```diff
@@ -13,7 +14,7 @@ public static class LetsEncryptSetupUtils
 {
         private const string AcmeClientUrl = "https://acme-v02.api.letsencrypt.org/directory";
 
-        public static async Task<byte[]> Setup(SetupInfo setupInfo,  SetupProgressAndResult progress, CancellationToken token)
+        public static async Task<byte[]> Setup(SetupInfo setupInfo,  SetupProgressAndResult progress, bool registerTcpDnsRecords, CancellationToken token)
```
Comment (gregolsky) [inline src/Raven.Server/Commercial/SetupWizard/LetsEncryptSetupUtils.cs:17]: I was wondering if adding it as optional to SetupInfo would not be better?

Comment (poissoncorp) [reply src/Raven.Server/Commercial/SetupWizard/LetsEncryptSetupUtils.cs:17]: https://issues.hibernatingrhinos.com/issue/RavenDB-19386/Move-bool-registerTcpDnsRecords-to-the-SetupInfo

File: tools/rvn/CommandLineApp.cs (hunk 5 above)
Comment (poissoncorp) [reply tools/rvn/CommandLineApp.cs:156]: it'll return `null` and the code will still work

**Review by gregolsky — APPROVED** [review]

File: test/SlowTests/Tools/SetupSecuredClusterUsingRvn.cs [hunk 9]
```diff
@@ -356,6 +356,7 @@ public async Task Should_Create_Secured_Cluster_From_Rvn_Using_Lets_Encrypt_Mode
             Processed = 0,
             Total = 4
         },
+            false,
```
Comment (ppekrol) [inline test/SlowTests/Tools/SetupSecuredClusterUsingRvn.cs:359]: use named parameters next time

**Pull request merged**